controllers/awscluster_controller.go

```diff
@@ -0,0 +1,279 @@
+	"github.com/go-logr/logr"
+
+	"github.com/giantswarm/capa-iam-operator/pkg/awsclient"
+	"github.com/giantswarm/capa-iam-operator/pkg/iam"
+	"github.com/giantswarm/capa-iam-operator/pkg/key"
+)
+
+const (
+	IRSASecretSuffix = "irsa-cloudfront"
+)
+
+// AWSClusterReconciler reconciles a Secret object
+type AWSClusterReconciler struct {
+	client.Client
+	EnableIRSARole            bool
+	Log                       logr.Logger
+	Scheme                    *runtime.Scheme
+	IAMClientAndRegionFactory func(awsclientgo.ConfigProvider) (iamiface.IAMAPI, string)
+	AWSClient                 awsclient.AwsClientInterface
+}
+
+// +kubebuilder:rbac:groups=core,resources=secrets,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=core,resources=secrets/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=core,resources=secrets/finalizers,verbs=update
+
+func (r *AWSClusterReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := r.Log.WithValues("namespace", req.Namespace, "AWSCluster", req.Name)
+
+	if !r.EnableIRSARole {
+		logger.Info("IRSA is not enabled")
+		return ctrl.Result{}, nil
+	}
+	logger.Info("Reconciling IRSA Secrets")
+
+	awsCluster := &capa.AWSCluster{}
+	err := r.Get(ctx, req.NamespacedName, awsCluster)
+	if err != nil {
+		logger.Error(err, "Failed to get the awsCLuster")
+		return ctrl.Result{}, err
+	}
+
+	var result ctrl.Result
+
+	if awsCluster.DeletionTimestamp == nil {
+		logger.Info("IRSA Roles - reconcile normal")
+		result, err = r.reconcileNormal(ctx, logger, awsCluster)
+	} else {
+		logger.Info("IRSA Roles - reconcile delete")
+		result, err = r.reconcileDelete(ctx, logger, awsCluster)
+	}
+
+	return result, err
+}
+
+func (r *AWSClusterReconciler) reconcileNormal(ctx context.Context, logger logr.Logger, awsCluster *capa.AWSCluster) (ctrl.Result, error) {
+	// add finalizer to AWSCluster
+	if !controllerutil.ContainsFinalizer(awsCluster, key.FinalizerName(iam.IRSARole)) {
+		patchHelper, err := patch.NewHelper(awsCluster, r.Client)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+		controllerutil.AddFinalizer(awsCluster, key.FinalizerName(iam.IRSARole))
+		err = patchHelper.Patch(ctx, awsCluster)
+		if err != nil {
+			logger.Error(err, "failed to add finalizer on Secret")
+			return ctrl.Result{}, err
+		}
+		logger.Info("successfully added finalizer to Secret", "finalizer_name", key.FinalizerName(iam.IRSARole))
+	}
+
+	cm := &corev1.ConfigMap{}
+	err := r.Get(
+		ctx,
+		types.NamespacedName{
+			Namespace: awsCluster.Namespace,
+			Name:      fmt.Sprintf("%s-%s", awsCluster.Name, "cluster-values"),
+		},
+		cm)
+	if err != nil {
+		logger.Error(err, "Failed to get the cluster-values configmap for cluster")
+		return ctrl.Result{}, err
+	}
+
+	if !controllerutil.ContainsFinalizer(cm, key.FinalizerName(iam.IRSARole)) {
+		patchHelper, err := patch.NewHelper(cm, r.Client)
+		if err != nil {
+			return ctrl.Result{}, err
```

Comment:
Using this function, we can get the stack trace when an error is returned, making it easier to know where the error happened
```
errors.WithStack(err)
```